File: examples/rtl_433_custom.py
#!/usr/bin/env python

"""Custom data handling example for rtl_433."""

# Start rtl_433 (rtl_433 -F syslog::1433), then this script

from __future__ import print_function

import socket
import json

# You can run rtl_433 and this script on different machines,
# start rtl_433 with `-F syslog:YOURTARGETIP:1433`, and change
# to `UDP_IP = "0.0.0.0"` (listen to the whole network) below.
UDP_IP = "127.0.0.1"
UDP_PORT = 1433


def parse_syslog(line):
    """Try to extract the payload from a syslog line."""
    line = line.decode("ascii")  # also UTF-8 if BOM
    if line.startswith("<"):
        # fields should be "<PRI>VER", timestamp, hostname, command, pid, mid, sdata, payload
        fields = line.split(None, 7)
        line = fields[-1]
    return line


def rtl_433_listen():
    """Listen to all messages in a loop forever."""
    # Open a UDP socket
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    # Bind the UDP socket to a listening address
    sock.bind((UDP_IP, UDP_PORT))

    # Loop forever
    while True:
        # Receive a message
        line, addr = sock.recvfrom(1024)

        try:
            # Parse the message format
            line = parse_syslog(line)
            # Decode the message as JSON
            data = json.loads(line)

            #
            # Change for your custom handling below, this is a simple example
            #
            label = data["model"]
            if "channel" in data:
                label += ".CH" + str(data["channel"])
            elif "id" in data:
                label += ".ID" + str(data["id"])

            # E.g. match `model` and `id` to a descriptive name.
            if data["model"] == "LaCrosse-TX" and data["id"] == 123:
                label = "Living Room"

            if "battery_ok" in data:
                if data["battery_ok"] == 0:
                    print(label + ' Battery empty!')

            if "temperature_C" in data:
                print(label + ' Temperature ', data["temperature_C"])

            if "humidity" in data:
                print(label + ' Humidity ', data["humidity"])

            # Ignore unknown message data and continue
        except KeyError:
            pass

        except ValueError:
            pass


if __name__ == "__main__":
    rtl_433_listen()

File: examples/rtl_433_custom.php
#!/usr/bin/php
<?php
/*
Short example of an UDP server written in PHP for rtl_433.

You need to have `php_cli` installed.

To run this as service use these systemd service files for
ubuntu, lubuntu, xubuntu, debian, ubuntu core and so on.

## Service file to start the rtl_433 process


[Unit]
Description=rtl433 udp service
After=network.target
[Service]
Restart=always
RestartSec=5
RemainAfterExit=no
User=root
ExecStart=/bin/sh -c "/usr/bin/rtl_433 -f 433920000 -f 433920000 -H 120 -F syslog:127.0.0.1:1433"

[Install]
WantedBy=multi-user.target


## Service file to start the php udp server


[Unit]
Description=syslog 433 udp service
After=network.target
StartLimitIntervalSec=0
[Service]
Type=simple
Restart=always
RestartSec=1
User=root
ExecStart=/usr/bin/php /home/user/rtl_433/examples/rtl_433_custom.php "0"

[Install]
WantedBy=multi-user.target
*/

error_reporting(E_ALL | E_STRICT);

//switch on debug mode
$debug = "1";
if (sizeof($argv) > 1)
{
    $debug  = $argv[1];
}
//udp server IP and Port for listen
$UDP_IP = "127.0.0.1";
$UDP_PORT = 1433;
//create socket and bind them
$socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
socket_bind($socket, $UDP_IP, $UDP_PORT);
//init of this variables
$from = '';
$port = 0;

//use the output of rtl_433 -f 433920000 -f 433920000 -H 120 -F syslog:127.0.0.1:1433"
//returns the json payload
function parse_syslog($line)
{
    //Try to extract the payload from a syslog line.//
    $line = mb_convert_encoding($line, "ASCII");

    if (startsWith($line,"<"))
    {
        //fields should be "VER", timestamp, hostname, command, pid, mid, sdata, payload
        $fields = explode(" ",$line, 8);
        $line = $fields[7];
    }
    return $line;
}

//server main loop
for (;;)
{
    //read from $socket into $line
    socket_recvfrom($socket, $line, 1024, 0, $from, $port);
    try
    {
        //parse $line -> returns the json payload
        $line = parse_syslog($line);

        /*
        do something with content of $line
        for example decode $line into a array
        $arr = json_decode($line,true);

        do something with that array and
        puted into a file as json

        file_put_contents('test.json', json_encode($arr, JSON_PRETTY_PRINT);
        */
    }
    catch (Exception $e) {
        echo "---------------------------------------------\n";
        echo 'Exception intercepted: ', $e->getMessage(), "\n";
        echo "------------------------------------------- -\n";
    }
}
?>

